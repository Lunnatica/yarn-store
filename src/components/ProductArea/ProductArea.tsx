import { useCallback, useEffect, useState } from 'react';

import { Product } from '../../types/types';
import { DetailModal } from '../DetailModal';
import { ProductList } from '../ProductList';
import { SearchBar } from '../SearchBar';

export type ProductListState = 'loading' | 'loaded' | 'error' | 'empty';

export const ProductArea: React.FC = () => {
  const [isOpen, setOpen] = useState(false);
  const [products, setProducts] = useState<Product[]>([]);
  const [selectedProduct, setSelectedProduct] = useState<Product>(null);
  const [productListState, setProductListState] = useState<ProductListState>('loading');

  const openModal = useCallback(
    (product: Product) => {
      setSelectedProduct(product);
      setOpen(true);
    },
    [setSelectedProduct, setOpen]
  );

  const fetchProducts = async (query?: string) => {
    try {
      setProductListState('loading');
      const response = await fetch(
        `${process.env.NEXT_PUBLIC_API_HOST}/products${query ? `?name_like=${query}` : ''}`
      );
      if (!response.ok) {
        setProductListState('error');
      }
      const data = await response.json();
      setProducts(data);
      if (data.length === 0) {
        setProductListState('empty');
      } else {
        setProductListState('loaded');
      }
    } catch (error) {
      setProductListState('error');
    }
  };

  useEffect(() => {
    fetchProducts();
  }, []);

  return (
    <>
      <SearchBar fetchProducts={fetchProducts} />

      <ProductList products={products} productListState={productListState} openModal={openModal} />

      {selectedProduct && (
        <DetailModal product={selectedProduct} isOpen={isOpen} onDismiss={() => setOpen(false)} />
      )}
    </>
  );
};
