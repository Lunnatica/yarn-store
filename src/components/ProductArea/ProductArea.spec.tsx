import { render, screen, waitFor } from '@testing-library/react';

import { ProductArea } from './ProductArea';
import { act } from 'react-dom/test-utils';

jest.mock('../DetailModal', () => ({
  DetailModal: () => <div data-testid="detail-modal" />,
}));
jest.mock('../ProductList', () => ({
  ProductList: () => <div data-testid="product-list" />,
}));
jest.mock('../SearchBar', () => ({
  SearchBar: () => <div data-testid="search-bar" />,
}));

describe('ProductArea', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.mockResponseOnce(JSON.stringify([]));
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should match the snapshot', async () => {
    let container = null;
    await act(async () => {
      ({ container } = render(<ProductArea />));
    });
    expect(container).toMatchSnapshot();
  });

  it('should call fetchProducts on mount', async () => {
    render(<ProductArea />);
    await waitFor(() => {
      expect(fetchMock).toHaveBeenCalledTimes(1);
    });
  });
});
