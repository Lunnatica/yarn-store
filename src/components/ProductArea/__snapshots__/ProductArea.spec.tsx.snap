// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductArea should match the snapshot 1`] = `
<div>
  <div
    data-testid="search-bar"
  />
  <div
    data-testid="product-list"
  />
</div>
`;
